storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java

```diff
@@ -126,4 +128,56 @@ private boolean isFinalized(final UnsignedLong slot) {
   public Optional<BeaconState> getNonfinalizedBlockState(final Bytes32 blockRoot) {
     return recentChainData.getBlockState(blockRoot);
   }
+
+  /**
+   * Returns the state which was proposed in or most recently before the requested slot on the chain
+   * specified by <code>headBlockRoot</code>. If the slot was empty, the state at the last filled
+   * slot is returned.
+   *
+   * @param slot the slot to get the effective block for
+   * @return the State at slot or the closest previous slot if empty
+   */
+  public SafeFuture<Optional<BeaconState>> getStateAtSlot(
+      final UnsignedLong slot, final Bytes32 headBlockRoot) {
+
+    final Store store = recentChainData.getStore();
+    if (store == null) {
+      LOG.trace("No state at slot {} because the store is not set", slot);
+      return STATE_NOT_AVAILABLE;
+    }
+
+    if (isFinalized(slot)) {
+      return historicalChainData.getFinalizedStateAtSlot(slot);
+    }
+
+    final BeaconState headState = store.getBlockState(headBlockRoot);
+    if (headState.getSlot().equals(slot)) {
+      return completedFuture(Optional.ofNullable(headState));
+    }
+
+    return completedFuture(recentChainData.getStateBySlot(slot));
+  }
+
+  public SafeFuture<Optional<BeaconState>> getStateAtBlock(final Bytes32 block) {
```

Comment:
`getStateAtBlock` probably isn't the right term - `getStateByBlockRoot(final Bytes32 blockRoot)` would be more what I'd expect.